Include/pystate.h

```diff
@@ -185,6 +204,74 @@ typedef struct _ts {
 } PyThreadState;
 #endif
 
+#ifndef Py_LIMITED_API
+
+typedef struct _frame *(*PyThreadFrameGetter)(PyThreadState *self_);
+
+struct _gilstate_globals {
+    int check_enabled;
+#ifdef Py_BUILD_CORE
+    /* Assuming the current thread holds the GIL, this is the
+       PyThreadState for the current thread. */
+    _Py_atomic_address tstate_current;
+#endif
+    PyThreadFrameGetter getframe;
+#ifdef WITH_THREAD
+    /* The single PyInterpreterState used by this process'
+       GILState implementation
+    */
+    /* TODO: Given interp_main, it may be possible to kill this ref */
+    PyInterpreterState *autoInterpreterState;
+    int autoTLSkey;
+#endif /* WITH_THREAD */
+};
+
+#include "warnings.h"
+#include "ceval.h"
+
+typedef struct pyruntimestate {
+    int initialized;
+    int core_initialized;
+    PyThreadState *finalizing;
+
+    struct pyinterpreters {
+#ifdef WITH_THREAD
+#ifdef Py_BUILD_CORE
```

Comment:
You *might* be able to manage a single flat struct by putting the private state at the end, but honestly, I think you'll have an easier time of it by starting with two structs (_PyRuntimeState and _PyRuntimePublicState), and then progressively reducing the public state to be as close to empty as possible, and instead requiring indirection through accessor and mutator functions.